### PR TITLE
[SYCL-MLIR][LoopInternalization] Get local memory pointer without `reqd_work_group_size`

### PR DESCRIPTION
**Pull request by whitneywhtsang**
When `reqd_work_group_size` is not given, query work group size with `nd_item.get_local_range(dim)`.

Continue from https://github.com/intel/llvm/pull/9767.

### CODE REVIEW

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 1]
```diff
@@ -44,6 +45,111 @@ using namespace mlir::polygeist;
 
 namespace {
 
+using ValueOrUnsignedTy = std::variant<Value, unsigned>;
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:48]: I don't see the point here to have a using declaration.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 2]
```diff
@@ -44,6 +45,111 @@ using namespace mlir::polygeist;
...
+};
+
+//===----------------------------------------------------------------------===//
+// WorkGroupSize
+//===----------------------------------------------------------------------===//
+
+/// Class to represent work group size of a kernel.
+class WorkGroupSize {
+public:
+  using ElemTy = ValueOrUnsignedTy;
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:110]: using ElemTy = std::variant<Value, unsigned>;

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 3]
```diff
@@ -44,6 +45,111 @@ using namespace mlir::polygeist;
 
 namespace {
 
+using ValueOrUnsignedTy = std::variant<Value, unsigned>;
+class ValueOrUnsigned {
+public:
+  /// Return a Value or unsigned of \p val depending on \p createVal.
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:51]: Given that this class is in a cpp file and the implementation of these member fuctions is simple I think you can remove the comments.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 4]
```diff
@@ -44,6 +45,111 @@ using namespace mlir::polygeist;
...
+    if (!createVal)
+      return val;
+    return builder.create<arith::ConstantIndexOp>(builder.getUnknownLoc(), val);
+  }
+
+  /// Return a Value of \p val. Create a Value if \p val is an unsigned.
+  static Value getValue(ValueOrUnsignedTy val, OpBuilder builder) {
+    if (std::holds_alternative<Value>(val))
+      return std::get<Value>(val);
+
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:63]: Remove

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp (hunk 1 above)
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:48]: removed.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp (hunk 3 above)
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:51]: removed.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp (hunk 4 above)
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:63]: removed.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp (hunk 2 above)
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:110]: done.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 5]
```diff
@@ -44,6 +45,111 @@ using namespace mlir::polygeist;
...
+private:
+  template <typename T, typename = std::enable_if_t<llvm::is_one_of<
+                            T, arith::AddIOp, arith::MulIOp>::value>>
+  static ValueOrUnsignedTy binaryOperator(const ValueOrUnsignedTy &lhs,
+                                          const ValueOrUnsignedTy &rhs,
+                                          OpBuilder builder) {
+    if (std::holds_alternative<unsigned>(lhs)) {
+      assert(std::holds_alternative<unsigned>(rhs));
+      if (std::is_same<T, arith::AddIOp>::value)
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:90]: is_same --> is_same_v

Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:90]: done.

**Review by etiotto — APPROVED** [review]

**Comment by Naghasan** [comment]
Looks ok to me, Ettore did all the nitpicks.

I'm wondering if we shouldn't try to introduce spec constant for this. However, this will most likely require a change in the runtime until we can rewrite part of the host.

**Comment by whitneywhtsang** [comment]
> Looks ok to me, Ettore did all the nitpicks.
> 
> I'm wondering if we shouldn't try to introduce spec constant for this. However, this will most likely require a change in the runtime until we can rewrite part of the host.

What is spec constant?

**Pull request merged**